src/main/java/io/jenkins/plugins/pipeline_log_fluentd_cloudwatch/LogStreamState.java

```diff
@@ -0,0 +1,505 @@
+
+        private void create(String logStreamName) throws IOException {
+            AWSLogs currentClient = client();
+            DescribeLogStreamsResult r = currentClient.describeLogStreams(new DescribeLogStreamsRequest(logGroupName).withLogStreamNamePrefix(logStreamName));
+            boolean found = false;
+            // TODO handle paging, in case we have a lot of similarly-named jobs
+            for (LogStream ls : r.getLogStreams()) {
+                if (ls.getLogStreamName().equals(logStreamName)) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                // First-time project.
+                LOGGER.log(Level.FINE, "Creating {0}", logStreamName);
+                currentClient.createLogStream(new CreateLogStreamRequest(logGroupName, logStreamName));
+            }
+        }
+
+        Auth authenticate() throws IOException {
+            AWSSecurityTokenServiceClientBuilder builder = AWSSecurityTokenServiceClientBuilder.standard();
+            CredentialsAwsGlobalConfiguration credentialsConfig = CredentialsAwsGlobalConfiguration.get();
+            String region = credentialsConfig.getRegion();
+            if (region != null) {
+                builder = builder.withRegion(region);
+            }
+            AmazonWebServicesCredentials jenkinsCredentials = credentialsConfig.getCredentials();
+            if (jenkinsCredentials != null) {
+                AWSStaticCredentialsProvider credentialsProvider = new AWSStaticCredentialsProvider(jenkinsCredentials.getCredentials());
+                builder.withCredentials(credentialsProvider);
+            }
+            AWSCredentialsProvider credentialsProvider = builder.getCredentials();
+            AWSCredentials masterCredentials = credentialsProvider != null ? credentialsProvider.getCredentials() : null;
+            String remotedAccessKeyId, remotedSecretAccessKey, remotedSessionToken;
+            String agentLogStreamName;
+            synchronized (agentLogStreamNames) {
+                for (int i = 1; ; i++) {
+                    String candidate = logStreamNameBase + "@agent" + i;
+                    if (agentLogStreamNames.add(candidate)) {
+                        agentLogStreamName = candidate;
+                        break;
+                    }
+                }
+            }
+            if (masterCredentials instanceof AWSSessionCredentials) {
+                // otherwise would just throw AWSSecurityTokenServiceException: Cannot call GetFederationToken with session credentials
+                String role = null;
+                if (jenkinsCredentials instanceof AWSCredentialsImpl) {
+                    role = Util.fixEmpty(((AWSCredentialsImpl) jenkinsCredentials).getIamRoleArn());
+                }
+                if (role != null) {
+                    // TODO would be cleaner if AmazonWebServicesCredentials had a getCredentials overload taking a policy
```

Comment:
Could we extract some of this logic into smaller, more specific auth chunks? Method is getting pretty long.

```java
Auth authenticate() throws IOException {
    if (masterCredentials instanceof AWSSessionCredentials) {
      ...
      Auth auth;
      if (role != null) {
        // TODO would be cleaner if Amaz......
        auth = authWithRole(region, role, agentLogStreamName);
      } else {
        auth = authWithtMasterCredentials(masterCredentials)
      } else if (masterCredentials == null) {
        LOGGER.log(Level.WARNING, "No AWS credentials to be found, giving up on limiting to a policy");
        auth = Auth(null, null, null, region, agentLogStreamName);
      } else {
        auth = authWithFederationToken(agentLogStreamName)
      }
      ....
      return auth;
    }
}
```